src/services/BaseApi.ts

```diff
@@ -0,0 +1,128 @@
+import axios, { AxiosInstance, AxiosRequestConfig, AxiosRequestHeaders, AxiosResponse } from 'axios'
+
+export type PrefectConfig = {
+  baseUrl: string,
+  token?: string,
+}
+
+export type GetApiBaseUrl<T extends PrefectConfig = PrefectConfig> = (config: T) => string
+export type GetApiHeaders<T extends PrefectConfig = PrefectConfig> = (config: T) => AxiosRequestHeaders
+
+export const getPrefectBaseUrl: GetApiBaseUrl = (config) => {
+  return config.baseUrl
+}
+
+export const getPrefectUIHeaders: GetApiHeaders = () => {
+  return {
+    'X-PREFECT-UI': true,
+  }
+}
+
+export const getAuthorizationHeaders: GetApiHeaders = (config) => {
+  const value: AxiosRequestHeaders = {}
+
+  if (config.token) {
+    value.Authorization = `bearer ${config.token}`
+  }
+
+  return value
+}
+
+export class BaseApi<T extends PrefectConfig = PrefectConfig> {
+  protected readonly apiConfig: T
+  protected routePrefix: string | undefined
+  protected getHeaders: GetApiHeaders[]
+  protected getBaseUrl: GetApiBaseUrl
+
+  public constructor(apiConfig: T) {
+    this.apiConfig = apiConfig
+
+    this.getHeaders = [getPrefectUIHeaders, getAuthorizationHeaders]
+    this.getBaseUrl = getPrefectBaseUrl
+  }
+
+  protected composeBaseUrl(): string {
+    return this.getBaseUrl(this.apiConfig)
+  }
+
+  protected composeHeaders(): AxiosRequestHeaders {
+    return this.getHeaders.reduce((headers, getHeaders) => {
+      return {
+        ...headers,
+        ...getHeaders(this.apiConfig),
+      }
+    }, {})
+  }
+
+  protected removeUndefinedPaths(input: string | undefined): input is string {
+    return !!input
+  }
```

Comment:
This is a better and more clear guard IMO. I'd also put this in the `variables.ts` utility file next to the `isNotNullish` utility.
```suggestion
  protected isDefined<T>(input: T | undefined): input is T {
    return input !== undefined
  }
```